From 1110aae8f3939f2828ee563ac338d2a4e56781af Mon Sep 17 00:00:00 2001
From: Nikolas Klauser <nikolasklauser@berlin.de>
Date: Fri, 2 Feb 2024 19:15:58 +0100
Subject: [PATCH 03/12] Implement __is_nothrow_convertible and use it in libc++
 (#80436)

GCC 13 has implemented this builtin.

(cherry picked from commit 9cc2122bf5a81f7063c2a32b2cb78c8d615578a1)
---
 .../llvm-project/clang/docs/LanguageExtensions.rst    |  1 +
 .../clang/include/clang/Basic/TokenKinds.def          |  1 +
 .../llvm-project/clang/lib/Sema/SemaExprCXX.cpp       | 11 +++++++++--
 3 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/interpreter/llvm-project/clang/docs/LanguageExtensions.rst b/interpreter/llvm-project/clang/docs/LanguageExtensions.rst
index 611bfbe4fa3..ba577cc3917 100644
--- a/interpreter/llvm-project/clang/docs/LanguageExtensions.rst
+++ b/interpreter/llvm-project/clang/docs/LanguageExtensions.rst
@@ -1569,6 +1569,7 @@ The following type trait primitives are supported by Clang. Those traits marked
 * ``__is_const`` (C++, Embarcadero)
 * ``__is_constructible`` (C++, MSVC 2013)
 * ``__is_convertible`` (C++, Embarcadero)
+* ``__is_nothrow_convertible`` (C++, GNU)
 * ``__is_convertible_to`` (Microsoft):
   Synonym for ``__is_convertible``.
 * ``__is_destructible`` (C++, MSVC 2013)
diff --git a/interpreter/llvm-project/clang/include/clang/Basic/TokenKinds.def b/interpreter/llvm-project/clang/include/clang/Basic/TokenKinds.def
index c10e2adfbe6..6cc6663578f 100644
--- a/interpreter/llvm-project/clang/include/clang/Basic/TokenKinds.def
+++ b/interpreter/llvm-project/clang/include/clang/Basic/TokenKinds.def
@@ -567,6 +567,7 @@ TYPE_TRAIT_1(__is_unsigned, IsUnsigned, KEYCXX)
 // Embarcadero Binary Type Traits
 TYPE_TRAIT_2(__is_same, IsSame, KEYCXX)
 TYPE_TRAIT_2(__is_convertible, IsConvertible, KEYCXX)
+TYPE_TRAIT_2(__is_nothrow_convertible, IsNothrowConvertible, KEYCXX)
 ARRAY_TYPE_TRAIT(__array_rank, ArrayRank, KEYCXX)
 ARRAY_TYPE_TRAIT(__array_extent, ArrayExtent, KEYCXX)
 // Name for GCC 6 compatibility.
diff --git a/interpreter/llvm-project/clang/lib/Sema/SemaExprCXX.cpp b/interpreter/llvm-project/clang/lib/Sema/SemaExprCXX.cpp
index c94b1e10950..ce4fd218d40 100644
--- a/interpreter/llvm-project/clang/lib/Sema/SemaExprCXX.cpp
+++ b/interpreter/llvm-project/clang/lib/Sema/SemaExprCXX.cpp
@@ -5780,7 +5780,8 @@ static bool EvaluateBinaryTypeTrait(Sema &Self, TypeTrait BTT, QualType LhsT,
     return Self.Context.typesAreCompatible(Lhs, Rhs);
   }
   case BTT_IsConvertible:
-  case BTT_IsConvertibleTo: {
+  case BTT_IsConvertibleTo:
+  case BTT_IsNothrowConvertible: {
     // C++0x [meta.rel]p4:
     //   Given the following function prototype:
     //
@@ -5841,7 +5842,13 @@ static bool EvaluateBinaryTypeTrait(Sema &Self, TypeTrait BTT, QualType LhsT,
       return false;
 
     ExprResult Result = Init.Perform(Self, To, Kind, FromPtr);
-    return !Result.isInvalid() && !SFINAE.hasErrorOccurred();
+    if (Result.isInvalid() || SFINAE.hasErrorOccurred())
+      return false;
+
+    if (BTT != BTT_IsNothrowConvertible)
+      return true;
+
+    return Self.canThrow(Result.get()) == CT_Cannot;
   }
 
   case BTT_IsAssignable:
-- 
2.43.0

